File: 316. Remove Duplicate Letters.py
# https://leetcode.com/problems/remove-duplicate-letters/

# Given a string s, remove duplicate letters so that every letter appears once
# and only once. You must make sure your result is the smallest in
# lexicographical order among all possible results.

# Note: This question is the same as 1081:
# https://leetcode.com/problems/smallest-subsequence-of-distinct-characters/

###############################################################################

# record last postion of each char
# use stack and pop previous chars when i) new char is smaller and ii) we can add the popped char back later

class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        last_pos = {}
        for idx, char in enumerate(s):
            last_pos[char] = idx
        
        stack = []
        for idx, char in enumerate(s):
            if char not in stack:
                # pop the previous chars if the new char is smaller
                # but only when we can add the popped char back: idx < last_pos[popped_char]
                while stack and char < stack[-1] and idx < last_pos[stack[-1]]:
                    stack.pop()
                stack.append(char)
        return ''.join(stack)

File: 316. Remove Duplicate Letters.cpp
// https://leetcode.com/problems/remove-duplicate-letters/

// Given a string s, remove duplicate letters so that every letter appears once and
// only once. You must make sure your result is the smallest in lexicographical
// order among all possible results.

////////////////////////////////////////////////////////////////////////////////

// sequentially add non-added char (use hashMap to record), pop prev char if
// i) new char is smaller and ii) can add back prev char later (use hashMap to record lastPos of each char)
class Solution {
public:
    string removeDuplicateLetters(string s) {
        int n = s.size();
        unordered_map<char, int> lastPos;
        unordered_map<char, bool> added;
        for (int i = 0; i < n; ++i) lastPos[s[i]] = i;
        
        string ans;
        for (int i = 0; i < n; ++i) {
            if (!added[s[i]]) {
                while (!ans.empty() and ans.back() > s[i] and lastPos[ans.back()] > i) {
                    added[ans.back()] = false; ans.pop_back();
                }
                ans += s[i]; added[s[i]] = true;
            }
        }
        return ans;
    }
};
